contracts/finance/VestingWalletConfidential.sol

```diff
@@ -0,0 +1,147 @@
+ * a beneficiary until a specified time.
+ *
+ * NOTE: Since the wallet is {Ownable}, and ownership can be transferred, it is possible to sell unvested tokens.
+ *
+ * NOTE: When using this contract with any token whose balance is adjusted automatically (i.e. a rebase token), make
+ * sure to account the supply/balance adjustment in the vesting schedule to ensure the vested amount is as intended.
+ */
+contract VestingWalletConfidential is Ownable {
+    error VestingWalletConfidentialInvalidDuration();
+    error VestingWalletConfidentialOnlyExecutor();
+
+    event ConfidentialFungibleTokenReleased(address indexed token, euint64 amount);
+
+    mapping(address token => euint64) private _confidentialFungibleTokenReleased;
+    uint64 private immutable _start;
+    uint64 private immutable _duration;
+    address private immutable _executor;
+
+    /**
+     * @dev Sets the beneficiary (owner), the start timestamp and the vesting duration (in seconds) of the vesting
+     * wallet.
+     */
+    constructor(
+        address executor_,
+        address beneficiary,
+        uint64 startTimestamp,
+        uint64 durationSeconds
+    ) Ownable(beneficiary) {
+        _start = startTimestamp;
+        _duration = durationSeconds;
+        _executor = executor_;
+    }
+
+    function executor() public view virtual returns (address) {
+        return _executor;
+    }
+
+    /**
+     * @dev Getter for the start timestamp.
+     */
+    function start() public view virtual returns (uint64) {
+        return _start;
+    }
+
+    /**
+     * @dev Getter for the vesting duration.
+     */
+    function duration() public view virtual returns (uint64) {
+        return _duration;
+    }
+
+    /**
+     * @dev Getter for the end timestamp.
+     */
+    function end() public view virtual returns (uint64) {
+        return start() + duration();
+    }
+
+    /**
+     * @dev Amount of token already released
+     */
+    function released(address token) public view virtual returns (euint64) {
+        return _confidentialFungibleTokenReleased[token];
+    }
+
+    /**
+     * @dev Getter for the amount of releasable `token` tokens. `token` should be the address of an
+     * {IConfidentialFungibleToken} contract.
+     */
+    function releasable(address token) public virtual returns (euint64) {
+        // vestedAmount >= released so this cannot overflow. released & vestedAmount can be 0 but are handled gracefully.
+        return FHE.sub(vestedAmount(token, uint64(block.timestamp)), released(token));
+    }
+
+    /**
+     * @dev Release the tokens that have already vested.
+     *
+     * Emits a {ConfidentialFungibleTokenReleased} event.
+     */
+    function release(address token) public virtual {
+        euint64 amount = releasable(token);
+        FHE.allowTransient(amount, token);
+        euint64 amountSent = IConfidentialFungibleToken(token).confidentialTransfer(owner(), amount);
+
+        // TODO: Could theoretically overflow
```

Comment:
How could it overflow considering `released + amount` couldn't be greater than token uint64 `supply`?

Reply:
Well it actually could be. Imagine you transfer 100% of the token supply (uint64.max) to the vesting contract, then release it. Then send it back to the vesting contract. Voila